harness/determined/core/_state_manager.py

```diff
@@ -0,0 +1,162 @@
+import atexit
+import logging
+import sys
+import threading
+import time
+from types import TracebackType
+from typing import Any, Optional, Union
+
+from determined.common import api
+from determined.common.api import bindings
+
+logger = logging.getLogger("determined.core")
+
+
+class StateManager:
```

Comment:
I challenge you to come up with a better name than "StateManager", which is only slightly more meaningful of a name than "ThingDoer".

Maybe: "CrashDetector"?  "ExitDetector"? Just "Heartbeat"?